### ISSUE TITLE
Reloadable rail broken +new development

### ISSUE THREAD

**Issue by L0ck0n** [PapaJoesSoup/BDArmory#747]
##### ISSUE TYPE

 - Bug Report   (Something is broken)

 - Enhancement  (Expand on existing features)

##### KSP and BDA Version

KSP 1.8.1

BDAC v1.3.2

##### SUMMARY
Reload in parts set up for reloadablerail module does not work in KSP. Turret works, missile launches, but pressing resupply button does nothing, new missile does not spawn.

+Some new things:
I know that you receive a lot of messages from people who want a lot of new things that are not possible to add into BDAc or KSP. I dont want to ask you for too much, but im creating mod, parts pack for KSP and BDAc with new missiles, launchers, turrets etc. 
Right now reloadablerail thing is pretty basic, guardmode is not able to use it (reload missile), you cant set number of reloads, or set reload time (spawn time for new missile), even in parts config file. Adding these things would create so many possibilities. Module reloadable rail would allow modders to create so many new sophisticated turrets and launchers.

##### STEPS TO REPRODUCE
Create basic turret in unity with proper gameobjects and other stuff necessary for BDAc turret and reloadable rail, export to game, fire missile and try to resupply missile.

**Comment by DoctorDavinci** [comment]
If this is regarding a part that you created then of course it will not work

The reloadable rail code requires specific part setup and transform naming in unity when setting up a model to be used in KSP

**Comment by L0ck0n** [comment]
Something like this?
MissileTransform, NodeB and other stuff for BDAc turret?
Everything is set up in Unity properly according to this:
https://i.imgur.com/MO1cK2d.png
And in game there is window with resupply button, but new missile does not spawn.

**Comment by SpannerMonkey** [comment]
@L0ck0n Yes that's all the correct info for building a reloadable system. Or would be if the system wasn't currently broken. 
Do note however that using this system in an actual turret ( a model containing pitch and yaw transforms) presents additional problems, missiles can only be loaded if the turret is in the 0 zero position, that is having all parts in the default rest position. If the turret is even slightly misplaced from this rest position, you will find the missile can clip into the actual turret, and will likely explode on launch.

@BDA team  I have tested this issue, with standard BDA parts, the reloadable mk41's upon which the dev was carried out, and on the raft of dev parts I threw together to test the limits of the system. **In all cases, the rail was non-functional**, all correct PAW items show as expected. 
Checking the log reveals an error generated every time the PAW resupply button is selected ( I have added this to the discord git page for ref) 
I also checked the Module Manager cfg cache, in order to examine the makeup of the reloadable rail, no module of the type "ModuleMissileRearm" is being applied to any part.

**Comment by L0ck0n** [comment]
@SpannerMonkey  @DoctorDavinci  @bdateam Thx, its strange, because some time ago (one or two updates ago) reloadablerail was working without problems in all configurations, with moving turrets, animated turrets (deploy animation) etc. 
Right now you can fire missile but cant reload a new one (example of a turret with deploy animation):
https://i.imgur.com/VT4GkhF.png
https://i.imgur.com/gBazMcc.png
https://i.imgur.com/8l6mdCj.png

Adding things like adjustable reload time, adjustable number of reloads or ability to set if guardmode can reload missiles or if reload is possible only by PAW button would be awesome too. 
Ability to set reload time could also prevent problems with reload when turret is not in zero, idle position (misplaced missile).

Anyway, reloadablerail thing adds so much to BDAc and game itself, there are so many turret configurations that could use this module.

**Comment by ChippedChap** [comment]
I looked into the resupply button doing nothing, and I believe I may have fixed it.

Using a KSP 1.9 debugging installation, I found that the resupply button, in the process of making a part, would call the constructor of ProtoPartSnapshot, which would throw a NullReferenceException.

I put a breakpoint on the constructor and when I looked at the Part object being passed, two properties, Landed and Splashed, also threw nullrefs when the debugger tried to read them. I also noticed that the parts vessel field was also null. To see if these three nullrefs (the one in game and the two from the properties) were related to the vessel field being null, I set the vessel field to a 'dummy' Vessel object before calling the constructor. Surprisingly, this worked and the missile was reloading correctly.

I attached pictures of my debugger before and after adding the single line that added the dummy vessel (grouped into 2 pdfs for convenience). You can see the fix [here on my fork](https://github.com/ChippedChap/BDArmory/commit/b5b447b9c8708a92fe85a435f96d42a0e5818aa2).

[dummy.pdf](https://github.com/PapaJoesSoup/BDArmory/files/4303964/dummy.pdf)
[nodummy.pdf](https://github.com/PapaJoesSoup/BDArmory/files/4303965/nodummy.pdf)

Not too big of a fan of this fix, the dummy vessel might cause unintended side effects, but it seems to work. The missiles fly to their targets as intended when fired. (I am using the reloadable rail part from ACA, but modified to use BDArmory's ModuleMissileRearm.) What's interesting is that ACA has exactly the same code but it works just fine when reloading through the intended indirect way (no PAW option to reload), but when recompiled to have the resupply method activatable by a PAW, I get the same nullref. It might be worth investigating into how the former case works to get a better solution.

Unrelated to the nullref fix, but I felt like this was important but wasn't made very clear. Reloading does not care about the position of a missile that has been moved/rotated with the transform tools. It only seems to care about the position (and orientation?) of the attachnode. This will affect any special offsets users might have or any rotating turrets present in the game.

Sorry if this has already been fixed, or isn't very high up the bugfixing priority. Based on this still being open I assumed no one has found a solution yet (which is surprising because it was only a single line fix.)

**Comment by ChippedChap** [comment]
After trying to fix an issue with a reload feature I was trying to add (reload bar for ModuleMissileManager with BDStagingAreaGauge) - I discovered the the dummy vessel caused issues. The solution breaks staging (the UI and the functionality - spacebar does nothing) completely.

I am working on a project to make ModuleMissileRearm more functional (adding new reload mechanics), but I'm quite worried it will get ignored, and the fact that there have been no replies for 8 days on this issue makes my concern worse.  Making a separate issue may make this less likely, but I want to make a working prototype build before making a feature request issue. Should I just make an issue right away?

Thank you for your time.
